### PR TITLE
Improve Microsoft.ReactNative compilation speed

### PR DESCRIPTION
**Pull request by vmoroz**
Previously we have reduced size of our PCH files to make sure that our code compiles on computers with low resources.
The issue is that it significantly increased compilation time on powerful computers.

In this change we improve the compilation speed by putting back includes into the pch.h file, but we make it conditional.
To improve the compilation speed developers can set `RNW_FASTBUILD` environment variable to `true`. It will result in bigger PCH file, but faster compilation speed.

Other related changes to improve the compilation speed:
- Added PCH to Folly and ReactCommon projects. We add `pch.h` to the `ForcedIncludeFiles`.
- Make sure that Microsoft.ReactNative always uses generated C++/WinRT files instead of occasionally picking them up from the SDK folder. While working on the PR I have noticed that the same C++/WinRT header file can be picked up by different .cpp files from different locations. We add the `$(GeneratedFilesDir)` to `AdditionalIncludeDirectories` to fix it.
- Make sure that we do not compile again RN files inside of Microsoft.ReactNative project: they are linked from the ReactCommon, and we disable their compilation in Microsoft.ReactNative.Cxx shared project. See the `BuildMSRNCxxReactCommon` property.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/7148)

### CODE REVIEW

File: vnext/Microsoft.ReactNative/Microsoft.ReactNative.vcxproj [hunk 1]
```diff
@@ -136,6 +138,7 @@
         PROJECT_ROOT_NAMESPACE=Microsoft::ReactNative;
         %(PreprocessorDefinitions)
       </PreprocessorDefinitions>
+      <PreprocessorDefinitions Condition="'$(RNW_FASTBUILD)' == 'true'">RNW_FASTBUILD;%(PreprocessorDefinitions)</PreprocessorDefinitions>
```
Comment (NickGerleman) [inline vnext/Microsoft.ReactNative/Microsoft.ReactNative.vcxproj:141]:
```suggestion
      <PreprocessorDefinitions Condition="'$(RnwFastBuild)' == 'true'">RNW_FASTBUILD;%(PreprocessorDefinitions)</PreprocessorDefinitions>
```

nit: MSBuild property name casing

**Review by NickGerleman — APPROVED** [review]
LGTM

File: vnext/Microsoft.ReactNative/Pch/pch.h [hunk 2]
```diff
@@ -26,6 +26,34 @@
...
+#include <winrt/Windows.System.h>
+#include <winrt/Windows.UI.Composition.h>
+#include <winrt/Windows.UI.ViewManagement.Core.h>
+#include <winrt/Windows.UI.ViewManagement.h>
+#include <winrt/Windows.UI.Xaml.Automation.Peers.h>
+#include <winrt/Windows.UI.Xaml.Controls.h>
+#include <winrt/Windows.UI.Xaml.Input.h>
+#include <winrt/Windows.UI.Xaml.Media.h>
+#include <winrt/Windows.UI.Xaml.Shapes.h>
+#include <winrt/Windows.UI.Xaml.h>
```
Comment (asklar) [inline vnext/Microsoft.ReactNative/Pch/pch.h:49]: can you include the namespace-agnostic headers instead?
i.e. UI.Composition.h, UI.Xaml.h, etc.

File: vnext/ReactCommon/ReactCommon.vcxproj [hunk 3]
```diff
@@ -163,9 +168,11 @@
       Using Strict FloatingPointModel on x64 to workaround a compiler issue, See https://github.com/microsoft/react-native-windows/issues/4122
       Using a patched copy of Yoga due to https://github.com/microsoft/react-native-windows/issues/3994
     -->
-    <ClCompile Condition="'$(Platform)' == 'x64'" Include=".\Yoga.cpp" AdditionalIncludeDirectories="$(YogaDir)\yoga;%(AdditionalIncludeDirectories)" DisableSpecificWarnings="4244;%(DisableSpecificWarnings)" FloatingPointModel="Strict" />
```
Comment (asklar) [inline vnext/ReactCommon/ReactCommon.vcxproj:166]: was this removal intentional? IIRC there was a floating point issue in x64 in Folly that this was added for

File: vnext/ReactCommon/ReactCommon.vcxproj [hunk 4]
```diff
@@ -163,9 +168,11 @@
...
-    <ClCompile Condition="'$(Platform)' != 'x64'" Include=".\Yoga.cpp" AdditionalIncludeDirectories="$(YogaDir)\yoga;%(AdditionalIncludeDirectories)" DisableSpecificWarnings="4244;%(DisableSpecificWarnings)" />
+    <ClCompile Include=".\Yoga.cpp" AdditionalIncludeDirectories="$(YogaDir)\yoga;%(AdditionalIncludeDirectories)" DisableSpecificWarnings="4244;%(DisableSpecificWarnings)" />
```
Comment (asklar) [inline vnext/ReactCommon/ReactCommon.vcxproj:171]: nit, can remove the .\ prefix

**Review by asklar — CHANGES_REQUESTED** [review]
the improvement makes sense to me - can you check out my comment about strict FP and winui3?

File: vnext/ReactCommon/ReactCommon.vcxproj (hunk 4 above)
Comment (vmoroz) [reply vnext/ReactCommon/ReactCommon.vcxproj]: I had to make the `<FloatingPointModel>Strict</FloatingPointModel>` for the whole project in order to use the PCH file. Otherwise it fails to compile. I had moved the related comment next to the property settings to indicate it.

File: vnext/ReactCommon/ReactCommon.vcxproj (hunk 3 above)
Comment (vmoroz) [reply vnext/ReactCommon/ReactCommon.vcxproj:166]: I had to make the `<FloatingPointModel>Strict</FloatingPointModel>` for the whole project in order to use the PCH file. Otherwise it fails to compile. I had moved the related comment next to the property settings to indicate it.

File: vnext/Microsoft.ReactNative/Pch/pch.h (hunk 2 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative/Pch/pch.h]: I have changed the code code use the namespace-agnostic headers where it is possible.

**Review by asklar — APPROVED** [review]
usually strictFP has runtime performance implications, so it'd be good if the perf hit can be limited to dev inner loop and not official builds

**Comment by vmoroz** [comment]
> usually strictFP has runtime performance implications, so it'd be good if the perf hit can be limited to dev inner loop and not official builds

My understanding is that we do it to work around a bug that we see at runtime. If we do not use it in the official builds, then the code will not work correctly.

**Review by jonthysell — APPROVED** [review]

File: vnext/ReactCommon/ReactCommon.vcxproj (hunk 3 above)
Comment (dannyvv) [reply vnext/ReactCommon/ReactCommon.vcxproj:166]: Please note that having unique properties (via attributes or child elements) per cpp file will sequentialize the compiles as msbuild is not multi-threaded inside a project and has to invoke a separate cl.exe call.

**Review by dannyvv — APPROVED** [review]
Will you do a future PR to update:
* build-rnw.yml and pass `/p:RNW_FASTBUILD=true`?
* msbuildtools.ts and pass '/p:RNW_FASTBUILD=true' when `highMemory` is set?

**Comment by vmoroz** [comment]
> Will you do a future PR to update:
> 
> * build-rnw.yml and pass `/p:RNW_FASTBUILD=true`?
> * msbuildtools.ts and pass `/p:RNW_FASTBUILD=true` when `highMemory` is set?

Yes, I have created issue #7158 to do that.
